### ISSUE TITLE
Switch statement over sealed class removed by Dart AOT compiler

### ISSUE THREAD

**Issue by MicheleQuasar** [dart-lang/sdk#56010]
I'm encountering an issue where the Dart AOT compiler removes non-dead code and get the following error at runtime:
`[ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Attempt to execute code removed by Dart AOT compiler (TFA)`

I have encountered this issue twice already but I haven't been able to build a minimal example, seems to break only in a relatively large codebase.

Both cases presented themselves on switch statements on the same sealed class. This is a `sealed class Super` with 7 child classes `final class ChildN extends Super` (for N=1, ..., 7). This is not an enum as some (but not all) of the child classes have some additional parameters. In both cases I had a `switch` statement on the class that is matching also the parameters of the children that have them, so something similar to:
```dart
String func(Super s) => switch (s) {
Child1() => "child1",
Child2() => "child2",
Child3() => "child3",
Child4() => "child4",
Child5() => "child5",
Child6(:final foo) => "child6$foo",
Child7(:final bar) => "child7$bar",
};
```
When debugging all would be fine, but when running the app in release mode I would get the error reported above. Changing the switch to a chain of if-else solves the issue, even though it is not desirable as I lose the benefit of the exhaustiveness check given by the switch statement.

Edit: Forgot to mention that the issue is not "deterministic". In the sense that the two switch statements I referred to above were both present at the same time in the codebase but the issue presented itself first on one of the instances and after some time on the other.

Dart version: 3.3.3
Platform: Windows

**Comment by alexmarkov** [comment]
This bug looks similar to https://github.com/dart-lang/sdk/issues/55310 which was fixed in 227244745ad6c3f6a25eb4812ec644986ce04346. Dart 3.3.3 doesn't have the fix.

@MicheleQuasar Could you reproduce the issue on the latest version of Dart SDK (3.5.0-40.0.dev or later)?
If the bug is still there, could you provide full and minimal example which triggers the bug?

**Comment by MicheleQuasar** [comment]
@alexmarkov I tried with the latest dev release of the SDK (3.5.0-180.3.beta) and it seems the latest instance of the bug I was experiencing is solved. 

Thinking a bit better at the situation, I probably noticed the second instance of the bug only at a later date as, until last week, that part of the code was not being run in release but only in development.

I'll set a reminder for when the 3.5.0 release comes out to revert back to a switch statement instead of a chain of ifs.
Should this particular issue come up again once 3.5.0 is released I'll try a bit harder to get a minimal example.

On my part I would consider the issue as closed, thanks!
